enterprise/server/remote_execution/containers/docker/docker_test.go

```diff
@@ -197,8 +197,8 @@ func TestDockerRun_Timeout_StdoutStderrStillVisible(t *testing.T) {
 	res := c.Run(ctx, cmd, workDir, container.PullCredentials{})
 
 	assert.True(
-		t, status.IsDeadlineExceededError(res.Error),
-		"expected DeadlineExceeded error, got: %s", res.Error)
+		t, status.IsUnavailableError(res.Error),
```

Comment:
I think the expected error code is DeadlineExceeded - if the container impl is returning Unavailable then that's probably a bug

Reply:
Hmm, I am not sure why we were initially expecting `DeadlineExceeded` since this is not a GRPC call. Docker client is REST over HTTP or UDS. 🤔 

The error I am getting here is

```
                Error:          Received unexpected error:
                                rpc error: code = Unavailable desc = container did not exit cleanly:
```

which correlates to the `Wait` call getting context canceled without a terminated container, which seems to be the correct behavior here?

Reply:
So this is the source of the Wait API

```
func (cli *Client) ContainerWait(ctx context.Context, containerID string, condition container.WaitCondition) (<-chan container.WaitResponse, <-chan error) {
	if versions.LessThan(cli.ClientVersion(), "1.30") {
		return cli.legacyContainerWait(ctx, containerID)
	}

	resultC := make(chan container.WaitResponse)
	errC := make(chan error, 1)

	query := url.Values{}
	if condition != "" {
		query.Set("condition", string(condition))
	}

	resp, err := cli.post(ctx, "/containers/"+containerID+"/wait", query, nil, nil)
	if err != nil {
		defer ensureReaderClosed(resp)
		errC <- err
		return resultC, errC
	}

	go func() {
		defer ensureReaderClosed(resp)

		body := resp.body
		responseText := bytes.NewBuffer(nil)
		stream := io.TeeReader(body, responseText)

		var res container.WaitResponse
		if err := json.NewDecoder(stream).Decode(&res); err != nil {
			// NOTE(nicks): The /wait API does not work well with HTTP proxies.
			// At any time, the proxy could cut off the response stream.
			//
			// But because the HTTP status has already been written, the proxy's
			// only option is to write a plaintext error message.
			//
			// If there's a JSON parsing error, read the real error message
			// off the body and send it to the client.
			_, _ = io.ReadAll(io.LimitReader(stream, containerWaitErrorMsgLimit))
			errC <- errors.New(responseText.String())  // <--- RETURNED HERE
			return
		}

		resultC <- res
	}()

	return resultC, errC
}
```

it seems like the error was returned by the json decoder failure?

Reply:
https://github.com/moby/moby/blob/6f2e0fac19b502e63aad68fd3d2d6678c4889022/client/container_wait.go#L60C8-L72

Reply:
Im thinking of doing this
```
		case err := <-errCh:
			if err.Error() == "" {
				err = context.DeadlineExceeded
			}
			return wrapDockerErr(err, "container did not exit cleanly")
```

But Im not sure if it's the right thing to do 🤔 
Perhaps the right fix here is to patch Docker CLI to return the json decode error? 🤔

Reply:
Ah sorry, I remembered that the error code here doesn't matter, as long as we just return some error in this case.

The executor always checks to see whether the context deadline exceeded, and if so, it overrides the command execution error to have DeadlineExceeded status required by the RBE protocol: https://github.com/buildbuddy-io/buildbuddy/blob/d0e9b2a25b12befdbe32577cbd06a324677fcb8e/enterprise/server/remote_execution/executor/executor.go#L368-L372

So I think returning Unavailable is fine :)